### ISSUE TITLE
Condition variable doesn't exist

### ISSUE THREAD

**Issue by jonbinney** [ros/ros_comm#369]
Inside of rospy's QueuedConnection, there is an exception handler which tries to grab a condition variable called "self._cond": https://github.com/ros/ros_comm/blob/indigo-devel/clients/rospy/src/rospy/impl/tcpros_pubsub.py#L431

But that class has no member variable called self._cond, it has only self._cond_data_available and self._cond_queue_swapped. As a result, I see the following error output:

```
Exception in thread Thread-23:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 504, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/opt/ros/hydro/lib/python2.7/dist-packages/rospy/impl/tcpros_pubsub.py", line 431, in _run
    with self._cond:
  File "/opt/ros/hydro/lib/python2.7/dist-packages/rospy/impl/tcpros_pubsub.py", line 390, in __getattr__
    return getattr(self._connection, name)
AttributeError: 'TCPROSTransport' object has no attribute '_cond'
```

**Comment by jonbinney** [comment]
@dirk-thomas this looks like it was added along with queue_size to handle disconnects

**Comment by dirk-thomas** [comment]
That looks definitely like a bug. I guess the line should be:

```
with self._lock:
```

Can you please describe what you did in order to reproduce the problem? And try of changing the referenced line with my guess fixed the problem for you?

**Comment by jonbinney** [comment]
I added queue_size to a rospy publisher. It may have had a client which disconnected which then triggered this. I'll try and create a small test case

**Comment by jonbinney** [comment]
Here's a script that will reproduce it:

```
import rospy
from std_msgs.msg import String

rospy.init_node('pub_disconnect')
pub = rospy.Publisher('foo', String, queue_size=100)

r = rospy.Rate(100)
while not rospy.is_shutdown():
    pub.publish(String('foooooo'))
    r.sleep()
```

Start that script, then start `rostopic echo foo` in a separate window. Then kill rostopic with a sigterm (ctrl-\) and you should see the publisher throw an exception.

**Comment by jonbinney** [comment]
Thanks @dirk-thomas !

**Comment by jonbinney** [comment]
Can you backport this fix to hydro-devel?

**Comment by jonbinney** [comment]
I'll make a pull-request to hydro-devel to backport it

**Comment by dirk-thomas** [comment]
Will be released as 1.10.2 into _hydro_ and 1.11.0 into _indigo_.

**Comment by jonbinney** [comment]
Ah, thanks!
